From 42a113038881b2d155d1c1ca253af01ecd87ebd7 Mon Sep 17 00:00:00 2001
From: Alessandro Astone <ales.astone@gmail.com>
Date: Fri, 29 Mar 2019 21:49:20 +0100
Subject: [PATCH 59/61] recovery: ignore refresh events while on browsing menus

 * Also avoid generating a refresh event when mounting
   a storage volume to browse it, it would end up being
   ignored anyways.

 Thanks to @tdm

Change-Id: Idabce4e90b484f34f83d310abdb345778cfc3bf0
---
 recovery.cpp | 12 ++++++++----
 ui.h         |  3 +--
 2 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/recovery.cpp b/recovery.cpp
index ec677e9a..6fb92467 100644
--- a/recovery.cpp
+++ b/recovery.cpp
@@ -735,7 +735,8 @@ int get_menu_selection(bool menu_is_main,
                        const char* const* headers,
                        const MenuItemVector& menu_items,
                        bool menu_only,
-                       int initial_selection, Device* device) {
+                       int initial_selection, Device* device,
+                       bool refreshable = false) {
   // Throw away keys pressed previously, so user doesn't accidentally trigger menu items.
   ui->FlushKeys();
 
@@ -800,7 +801,9 @@ int get_menu_selection(bool menu_is_main,
           chosen_item = Device::kGoHome;
           break;
         case Device::kRefresh:
-          chosen_item = Device::kRefresh;
+          if (refreshable) {
+            chosen_item = Device::kRefresh;
+          }
           break;
       }
     } else if (!menu_only) {
@@ -869,7 +872,7 @@ static std::string browse_directory(const std::string& path, Device* device) {
       return "";
     }
     if (chosen_item == Device::kRefresh) {
-      return "@refresh";
+      continue;
     }
 
     const std::string& item = zips[chosen_item];
@@ -1225,6 +1228,7 @@ static int apply_from_storage(Device* device, VolumeInfo& vi, bool* wipe_cache)
     if (!VolumeManager::Instance()->volumeMount(vi.mId)) {
         return INSTALL_ERROR;
     }
+    ui->VolumesChanged();
 
     std::string path;
     do {
@@ -1289,7 +1293,7 @@ refresh:
     int status = INSTALL_ERROR;
 
     int chosen = get_menu_selection(false, MT_LIST, headers, items,
-                                    false, 0, device);
+                                    false, 0, device, true/*refreshable*/);
     if (chosen == Device::kRefresh) {
         goto refresh;
     }
diff --git a/ui.h b/ui.h
index 5ea54d83..7144c5d9 100644
--- a/ui.h
+++ b/ui.h
@@ -245,6 +245,7 @@ class RecoveryUI {
 
   // Notify of volume state change
   void onVolumeChanged() { volumes_changed_ = 1; }
+  bool VolumesChanged();
 
   virtual bool MenuShowing() const = 0;
   virtual bool MenuScrollable() const = 0;
@@ -297,8 +298,6 @@ class RecoveryUI {
 
   bool IsUsbConnected();
 
-  bool VolumesChanged();
-
   static void* time_key_helper(void* cookie);
   void time_key(int key_code, int count);
 
-- 
2.17.1

